### PR TITLE
[mesh] Rendering of scalar and vector datasets

### PR DESCRIPTION
**Pull request by wonder-sk**
Introducing support for rendering of scalar and vector datasets for mesh layers!

The goal of this PR is to have at usable rendering core of data in mesh layers for QGIS 3.2. There is no GUI support yet, the datasets must be added and styled via Python API.

Internal copy of the MDAL library is updated to version 0.0.3 which adds API for reading of datasets and adds support for reading text and binary .dat files used to store scalar/vector datasets.

Rendering of scalar data on nodes - interpolated:

interpolated

Showcase of rendering configuration (mesh rendering, scalar data rendering, vector data rendering):

mesh_datasets

Code used in the above picture:

```
base = '~/mesh/MDAL/tests/data/'
layer = QgsMeshLayer(base+'2dm/quad_and_triangle.2dm', 'q_and_t', 'mdal')
layer.dataProvider().addDataset(base+'ascii_dat/quad_and_triangle_els_vector.dat')
QgsProject.instance().addMapLayer(layer)

rs = layer.rendererTriangularMeshSettings()
rs.setEnabled(True)
rs.setColor(QColor(255, 0, 255))
layer.setRendererTriangularMeshSettings(rs)

rs.setEnabled(False)
layer.setRendererTriangularMeshSettings(rs)

layer.setActiveScalarDataset(0)
layer.triggerRepaint()

layer.setActiveScalarDataset(-1)
layer.triggerRepaint()

layer.setActiveVectorDataset(0)
layer.triggerRepaint()

rs = layer.rendererVectorSettings()
rs.setShaftLengthMethod(QgsMeshRendererVectorSettings.Scaled)
rs.setScaleFactor(100)
layer.setRendererVectorSettings(rs)
```

### CODE REVIEW

**Comment by nyalldawson** [comment]
Can we get a feature freeze exemption for this? There's a lot to review in a day :fearful:

**Comment by wonder-sk** [comment]
> Can we get a feature freeze exemption for this? There's a lot to review in a day

Sure, I do not have a problem with that. Also, it does not really affect existing QGIS functionality...

By the way, about 1/3 of the additions are coming from MDAL (drivers for two formats, utilities and API additions) that probably do not need reviewing as technically that code is not a part of QGIS project...

This PR is probably the last large PR one for mesh layer support, the upcoming PRs will be incremental improvements of the existing mesh layer code (e.g. spatial indexing, caching, identify support, GUI support, ...)

**Comment by nyalldawson** [comment]
> This PR is probably the last large PR one for mesh layer support, the upcoming PRs will be incremental improvements of the existing mesh layer code (e.g. spatial indexing, caching, identify support, GUI support, ...)

Are they also targeted for 3.2?

**Comment by wonder-sk** [comment]
> Are they also targeted for 3.2?

Not really, for 3.2 there will be only bug fix PRs for mesh layer code...

File: src/core/mesh/qgsmeshlayerrenderer.cpp [hunk 1]
```diff
@@ -91,3 +197,52 @@ void QgsMeshLayerRenderer::renderMesh( const std::unique_ptr<QgsSymbol> &symbol,
...
+  lst << QgsColorRampShader::ColorRampItem( vMin, mRendererScalarSettings.minColor(), QString::number( vMin ) );
+  lst << QgsColorRampShader::ColorRampItem( vMax, mRendererScalarSettings.maxColor(), QString::number( vMax ) );
+
+  QgsColorRampShader *fcn = new QgsColorRampShader( vMin, vMax );
+  fcn->setColorRampItemList( lst );
+  QgsRasterShader *sh = new QgsRasterShader( vMin, vMax );
+  sh->setRasterShaderFunction( fcn );  // takes ownership of fcn
+  QgsMeshLayerInterpolator interpolator( mTriangularMesh, mScalarDatasetValues, mScalarDataOnVertices, mContext, mOutputSize );
+  QgsSingleBandPseudoColorRenderer r( &interpolator, 0, sh );  // takes ownership of sh
+  QgsRasterBlock *bl = r.block( 0, mContext.extent(), mOutputSize.width(), mOutputSize.height(), mFeedback.get() );
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshlayerrenderer.cpp]: unique_ptr

File: src/core/mesh/qgsmeshrenderersettings.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+{
+  mFilterMax = vectorFilterMax;
+}
+
+QgsMeshRendererVectorSettings::ArrowScalingMethod QgsMeshRendererVectorSettings::shaftLengthMethod() const
+{
+  return mShaftLengthMethod;
+}
+
+void QgsMeshRendererVectorSettings::setShaftLengthMethod( const QgsMeshRendererVectorSettings::ArrowScalingMethod &shaftLengthMethod )
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshrenderersettings.cpp]: pass by value, not const ref here

File: src/core/mesh/qgsmeshvectorrenderer.cpp [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+  mMaxMag = *std::max_element( mDatasetValuesMag.constBegin(), mDatasetValuesMag.constEnd() );
+}
+
+QgsMeshVectorRenderer::~QgsMeshVectorRenderer() = default;
+
+void QgsMeshVectorRenderer::draw()
+{
+  // Set up the render configuration options
+  QPainter *p = mContext.painter();
+  p->setRenderHint( QPainter::Antialiasing );
```
Comment (nyalldawson) [inline src/core/mesh/qgsmeshvectorrenderer.cpp]: Only set antialiasing if the render context has that flag present

File: src/core/mesh/qgstriangularmesh.cpp [hunk 4]
```diff
@@ -79,6 +125,23 @@ void QgsTriangularMesh::update( QgsMesh *nativeMesh, QgsRenderContext *context )
     }
   }
 
+  // CALCULATE CENTROIDS
+  mNativeMeshFaceCentroids.resize( nativeMesh->faces.size() );
+  for ( int i = 0; i < nativeMesh->faces.size(); ++i )
+  {
+    const QgsMeshFace &face = nativeMesh->faces.at( i ) ;
+    QVector<QPointF> points;
```
Comment (nyalldawson) [inline src/core/mesh/qgstriangularmesh.cpp]: add

    points.reserve( face.size() );

File: src/providers/mdal/qgsmdalprovider.cpp [hunk 5]
```diff
@@ -93,6 +96,89 @@ QgsMeshFace QgsMdalProvider::face( int index ) const
...
+  }
+
+  return val;
+}
+
+void QgsMdalProvider::refreshDatasets()
+{
+  mDatasets.clear();
+  int n = MDAL_M_datasetCount( mMeshH );
+  for ( int i = 0; i < n; ++i )
```
Comment (nyalldawson) [inline src/providers/mdal/qgsmdalprovider.cpp]: mDatasets.resize( 0 ); // keeps allocated space - potentially avoids reallocation
mDatasets.reserve(n);

**Comment by nyalldawson** [comment]
Note regarding the review comments: I haven't really looked in to the mesh rendering specifics, just the general code. Also there's some changes I'd ideally make to the API, but given that it's not fixed there's really no harm in iteratively adapting the API as required. :)

File: src/core/mesh/qgstriangularmesh.cpp (hunk 4 above)
Comment (PeterPetrik) [reply src/core/mesh/qgstriangularmesh.cpp]: added

File: src/providers/mdal/qgsmdalprovider.cpp (hunk 5 above)
Comment (PeterPetrik) [reply src/providers/mdal/qgsmdalprovider.cpp]: added

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+ *
+ * \since QGIS 3.2
+ */
+class CORE_EXPORT QgsMeshRendererScalarSettings
+{
+  public:
+    //! Returns color representing maximum scalar value in the dataset
+    QColor maxColor() const;
+    //! Sets color representing maximum scalar value in the dataset
+    void setMaxColor( const QColor &maxColor );
```
[the remark this answers is not in the document]
Comment (PeterPetrik) [reply src/core/mesh/qgsmeshrenderersettings.h]: sure, that is the plan

**Comment by PeterPetrik** [comment]
@nyalldawson thanks a lot for a review. I tried to address all the points. I think the API changes when we start working on the GUI widgets for styling.

**Comment by wonder-sk** [comment]
If there are no objections I will merge this later today...

File: python/core/auto_generated/mesh/qgsmeshdataprovider.sip.in [hunk 7]
```diff
@@ -10,64 +10,255 @@
...
+%End
+
+    QgsMeshDatasetMetadata( bool isScalar,
+                            bool isValid,
+                            bool isOnVertices,
+                            const QMap<QString, QString> &extraOptions );
+%Docstring
+Constructs a valid metadata object
+
+:param isScalar: dataset contains scalar data, particularly the y-value of QgsMeshDatasetValue is NaN
```
Comment (nyalldawson) [inline python/core/auto_generated/mesh/qgsmeshdataprovider.sip.in]: Particularly -> specifically

**Comment by nyalldawson** [comment]
Just a tiny doc change

**Pull request merged**